src/nvim/lua/executor.c

```diff
@@ -1264,6 +1264,59 @@ static void nlua_add_treesitter(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
   lua_setfield(lstate, -2, "_ts_parse_query");
 }
 
+int nlua_expand_pat(char_u *pat, int *num_results, char_u ***results)
+{
+  lua_State *const lstate = nlua_enter();
+
+  // [ vim ]
+  lua_getglobal(lstate, "vim");
+
+  // [ vim, vim._expand_pat ]
+  lua_getfield(lstate, -1, "_expand_pat");
+  luaL_checktype(lstate, -1, LUA_TFUNCTION);
+
+  // [ vim, vim._log_keystroke, buf ]
+  lua_pushlstring(lstate, (const char *)pat, STRLEN(pat));
+
+  if (lua_pcall(lstate, 1, 1, 0)) {
+    nlua_error(
+        lstate,
+        _("Error executing vim._expand_pat: %.*s"));
+  }
+
+  Error err = ERROR_INIT;
+
+  *num_results = 0;
+  *results = NULL;
+
+  Array completions = nlua_pop_Array(lstate, &err);
+  if (ERROR_SET(&err)) {
+    return FAIL;
+  }
+
+  garray_T result_array;
+  ga_init(&result_array, (int)sizeof(char *), 80);
+  for (size_t i = 0; i < completions.size; i++) {
+    Object v = completions.items[i];
+
+    if (v.type != kObjectTypeString) {
+      goto cleanup;
+    }
+
+    GA_APPEND(char_u *, &result_array, (char_u *)v.data.string.data);
+  }
+
+  *results = result_array.ga_data;
+  *num_results = result_array.ga_len;
+
+  return OK;
+
+cleanup:
+  ga_clear(&result_array);
```

Comment:
do you even `api_free_array(completions)` ?

Reply:
Do I need it for both cases (i.e. with an error and w/out?)

Reply:
Yes.

Reply:
:+1: OK, sounds good. 

the `ga_*` stuff gets cleared automatically by the parent completion stuff, right?

Reply:
I think so, but a complete integration test is gonna confirm that on CI :)